sklearn/metrics/_regression.py

```diff
@@ -791,25 +856,50 @@ def r2_score(y_true, y_pred, *, sample_weight=None, multioutput="uniform_average
     denominator = (
         weight * (y_true - np.average(y_true, axis=0, weights=sample_weight)) ** 2
     ).sum(axis=0, dtype=np.float64)
-    nonzero_denominator = denominator != 0
-    nonzero_numerator = numerator != 0
-    valid_score = nonzero_denominator & nonzero_numerator
-    output_scores = np.ones([y_true.shape[1]])
-    output_scores[valid_score] = 1 - (numerator[valid_score] / denominator[valid_score])
-    # arbitrary set to zero to avoid -inf scores, having a constant
-    # y_true is not interesting for scoring a regression anyway
-    output_scores[nonzero_numerator & ~nonzero_denominator] = 0.0
+
+    return _assemble_r2_explained_variance(
+        numerator=numerator,
+        denominator=denominator,
+        nb_outputs=y_true.shape[1],
+        multioutput=multioutput,
+        force_finite=force_finite,
+    )
+
+
+def _assemble_r2_explained_variance(
```

Comment:
Could we move this function above the `explained_variance_score`? It might be more intuitive to define the function before it is called.

Reply:
indeed